Fix backwards compatibility

patrick91 opened the pull request:
https://github.com/strawberry-graphql/strawberry/pull/2037 removed `get_result` from StrawberryField, I didn't realise there are libraries relying on it (strawberry-django-plus being one of them).

I'm raising this PR to discuss how we could keep backwards compatibility with the change I made. 

The `if hasattr(field, "get_result"):` check is not ideal as it is being called on every resolver and (I think) it is more expensive than a `if field.some_property`.

/cc @bellini666 do you think strawberry-django-plus could migrate away from `get_result`?

bellini666 commented:
Hi @patrick91 ,

It is not only `strawberry-django-plus` but the [`strawberry-graphql-django` itself](https://github.com/strawberry-graphql/strawberry-graphql-django/blob/main/strawberry_django/fields/field.py#L101) (I just improved how it worked, specially regarding sync/async stuff in my lib).

We could try to move away from `get_result`, but specially in django there are some things that a `getattr` would not allow to properly resolve, like model managers (as you can see on `strawberry-graphql-django` itself). In `strawberry-django-plus`'s cases, there are also some optimizations done there and also some attempts on optimize inner queries to avoid n+1 problems in the get_result too.

But I can see 2 solutions to improve performance:

1) Instead of using `hasattr` with `field.get_result`, which results in 2 lookups for the attribute in cases that `get_result` is really there, we could use a try/except AttributeError instead (which in newer python versions has even zero cost).

e.g.

```python
try:
    return field.get_result(_source, info=info, args=field_args, kwargs=field_kwargs)
except AttributeError:
    return self.config.default_resolver(_source, field.python_name)
```

2) Even better yet, since `_get_result` is defined once, we can create those conditionally based on the existence of a `base_resolver`, `get_result` or fallback to `default_resolver`. E.g.:

```python
if base_resolver := field.base_resolver:
    def _get_result(_source: Any, info: Info, **kwargs):
        field_args, field_kwargs = _get_arguments(source=_source, info=info, kwargs=kwargs)
        return base_resolver(*field_args, **field_kwargs)
elif get_result := getattr(field, 'get_result', None):
    def _get_result(_source: Any, info: Info, **kwargs):
        field_args, field_kwargs = _get_arguments(source=_source, info=info, kwargs=kwargs)
        return get_result(_source, info=info, args=field_args, kwargs=field_kwargs)
else:
    resolver = self.config.default_resolver
    python_name = field.python_name
    def _get_result(_source: Any, info: Info, **kwargs):
        return default_resolver(_source, python_name)
```

With this, we even cache other attribute accesses like `self.config.default_resolver` and `field.python_name`, reducing some overhead even more (although minimum in this case).

For typing purposes, we can even type `get_result` as an `Optional` attribute in `StrawberryField` so that anyone implementing it has to implement the proper protocol for it.

What do you think?

patrick91 commented:
> Hi @patrick91 ,

> We could try to move away from `get_result`, but specially in django there are some things that a `getattr` would not allow to properly resolve, like model managers (as you can see on `strawberry-graphql-django` itself). In `strawberry-django-plus`'s cases, there are also some optimizations done there and also some attempts on optimize inner queries to avoid n+1 problems in the get_result too.
> 
> But I can see 2 solutions to improve performance:
> 
> 1. Instead of using `hasattr` with `field.get_result`, which results in 2 lookups for the attribute in cases that `get_result` is really there, we could use a try/except AttributeError instead (which in newer python versions has even zero cost).
> 
> e.g.
> 
> ```python
> try:
>     return field.get_result(_source, info=info, args=field_args, kwargs=field_kwargs)
> except AttributeError:
>     return self.config.default_resolver(_source, field.python_name)
> ```
> 
> 2. Even better yet, since `_get_result` is defined once, we can create those conditionally based on the existence of a `base_resolver`, `get_result` or fallback to `default_resolver`. E.g.:
> 
> ```python
> if get_result := getattr(field, 'get_result', None):
>     def _get_result(_source: Any, info: Info, **kwargs):
>         field_args, field_kwargs = _get_arguments(source=_source, info=info, kwargs=kwargs)
>         return get_result(_source, info=info, args=field_args, kwargs=field_kwargs)
> elif base_resolver := field.base_resolver:
>     def _get_result(_source: Any, info: Info, **kwargs):
>         field_args, field_kwargs = _get_arguments(source=_source, info=info, kwargs=kwargs)
>         return base_resolver(*field_args, **field_kwargs)
> else:
>     field_resolver = self.config.default_resolver
>     python_name = field.python_name
>     def _get_result(_source: Any, info: Info, **kwargs):
>         return field_resolver(_source, python_name)
> ```
> 
> With this, we even cache other attribute accesses like `self.config.default_resolver` and `field.python_name`, reducing some overhead even more (although minimum in this case).
> 
> For typing purposes, we can even type `get_result` as an `Optional` attribute in `StrawberryField` so that anyone implementing it has to implement the proper protocol for it.
> 
> What do you think?

I think this might be the way to, we keep `get_result` as optional and use that if it is defined. 

By the way, we do have a PR that tries to optimise things here: https://github.com/strawberry-graphql/strawberry/pull/1853 

we could implement the try/catch for now and improve things later or we can take this approach too: https://github.com/strawberry-graphql/strawberry/pull/1949

bellini666 commented:
Hey @patrick91 ,

Sure, lets do that then, and I'll suggest this optimization at that other PR later :)

Sad that there's no easy way to remove `get_result`... I'll keep thinking of a way for doing that, but right now there's no easy way around it :(

patrick91 commented:
ok, I've changed my mind a bit, I've updated the code to pass the default resolver to `StrawberryField`. I still don't think this is ideal, but it seems like the best solution for the time being /cc @BryceBeagle (we chatted a bit about this)

I think the next things we should do after merging this PR are:

1. see if we can customise `base_resolver` in strawerry django (and django plus) instead of using `get_resolver`
2. if that works maybe we could rename `base_resolver` to `resolver`
3. move most of the logic done inside `from_resolver` somewhere else, there are things that aren't really related to schema conversion (checking permissions and getting the result)

<img width="663" alt="CleanShot 2022-07-26 at 12 25 28@2x" src="https://user-images.githubusercontent.com/667029/180994920-391da1cd-da51-40df-b3e7-196bf3693e4f.png">

patrick91 commented:
/pre-release

botberry commented:
# Pre-release
:wave:

Releasing commit [12b242532159045d97cdabda347f8d2ce1f63836] to PyPi as pre-release! :package:

botberry commented:
# Pre-release
:wave:

Pre-release **0.121.1.dev.1658834895** [12b242532159045d97cdabda347f8d2ce1f63836] has been released on PyPi! :rocket:
You can try it by doing:

```shell
poetry add strawberry-graphql==0.121.1.dev.1658834895
```

botberry commented:
Thanks for adding the `RELEASE.md` file!

![](https://media.giphy.com/media/xq1FxHkABwW7m/giphy.gif)

Here's a preview of the changelog:

---

This release adds a backward compatibility layer with libraries
that specify a custom `get_result`.


---
Here's the preview release card for twitter:

![](https://i.imgur.com/hYO7cHP.png)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @patrick91 for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```
<!-- action-check: release-file -->

bellini666 commented on strawberry/schema/schema_converter.py:
```diff
@@ -483,10 +483,11 @@ def _get_result(_source: Any, info: Info, **kwargs):
 
-            if field.base_resolver:
-                return field.base_resolver(*field_args, **field_kwargs)
+            field.default_resolver = self.config.default_resolver  # type: ignore
```
Does this need to be set here or could it be set outside of `_get_result`? If I understand this code correctly, when defining this function we already have the config set, and it isn't something that should be changed at runtime

patrick91 replied:
yes, I can move that 😊 will do that later 😊

bellini666 commented:
> ok, I've changed my mind a bit, I've updated the code to pass the default resolver to `StrawberryField`. I still don't think this is ideal, but it seems like the best solution for the time being /cc @BryceBeagle (we chatted a bit about this)

I agree! The major issue I see here is to allow fields to resolve themselves in a custom way for complex things like the django integration stuff, while keeping simple fields (which uses `getattr`/`getitem` directly) with a minimum overhead as possible.

I don't know if there are other integrations that relies on `get_result` to handle those complex cases like django requires, but probably there will be such cases too.

> 
> I think the next things we should do after merging this PR are:
> 
> 1. see if we can customise `base_resolver` in strawerry django (and django plus) instead of using `get_resolver`

You mean defining a `base_resolver` for those strawberry django fields that would be used if no custom resolver is passed to the field? I think this can work great!

I can do these changes myself on both libs and then we can come back here and really remove `get_result` :)

> 2. if that works maybe we could rename `base_resolver` to `resolver`

👍🏼

bellini666 reviewed (approved):
The change is now very simple and it seems fine to me overall!

I've made a small suggestion for a small performance improvement, but it is up to you to use it or not.

Apakottur commented:
I'm a bit late to the party but I'll just add that in my team we are also overriding `StrawberryField.get_result` to add a context manager around the execution of the resolver (to add logging, error management and other asbtractions that are common to all resolvers).

We tried using [Extensions](https://strawberry.rocks/docs/guides/custom-extensions) but they are not as powerful as being able to run code directly before/after/around the call to the resolver function (which is what we get by overriding `get_result`).

If it matters, we're not using Django.

+100 for this fix :medal_sports:

BryceBeagle reviewed (changes requested):
Glad to see the logic being moved out of the schema converter and back to the field!

> I think the next things we should do after merging this PR are:
> 
> 1. see if we can customise `base_resolver` in strawerry django (and django plus) instead of using `get_resolver`

Can you explain what you mean here? What would customization look like beyond overriding a method?

> 2. if that works maybe we could rename `base_resolver` to `resolver`

I think this is a good idea. I can't remember why I went for `base_resolver` instead of `resolver` originally. 

> 3. move most of the logic done inside `from_resolver` somewhere else, there are things that aren't really related to schema conversion (checking permissions and getting the result)

Completely agree here. The schema converter shouldn't hold any of the runtime logic. It should just be for creating the datastructures to pass to `graphql-core`

patrick91 commented on strawberry/field.py:
```diff
@@ -43,6 +45,7 @@
 
 class StrawberryField(dataclasses.Field):
     python_name: str
+    default_resolver: Callable[[Any, str], object] = getattr
```
we currently set it from in the schema converter

patrick91 commented on strawberry/field.py:
```diff
@@ -144,6 +147,19 @@ def __call__(self, resolver: _RESOLVER_TYPE) -> "StrawberryField":
+    ) -> Union[Awaitable[Any], Any]:
+        """
+        Calls the resolver defined for the StrawberryField. If the field doesn't have a
+        resolver defined we default to using getattr on `source`.
```
updated 😊

patrick91 commented:
> Glad to see the logic being moved out of the schema converter and back to the field!
> 
> > I think the next things we should do after merging this PR are:
> > 
> > 1. see if we can customise `base_resolver` in strawerry django (and django plus) instead of using `get_resolver`
> 
> Can you explain what you mean here? What would customization look like beyond overriding a method?

see: https://github.com/strawberry-graphql/strawberry-graphql-django/blob/main/strawberry_django/fields/field.py#L101-L10

I think most of this logic should go a way in favour of a `resolver` property, but I think it might better to discuss that on a PR with some code 😊

> 
> > 2. if that works maybe we could rename `base_resolver` to `resolver`
> 
> I think this is a good idea. I can't remember why I went for `base_resolver` instead of `resolver` originally.
> 
> > 3. move most of the logic done inside `from_resolver` somewhere else, there are things that aren't really related to schema conversion (checking permissions and getting the result)
> 
> Completely agree here. The schema converter shouldn't hold any of the runtime logic. It should just be for creating the datastructures to pass to `graphql-core`
